src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -18955,5 +19045,160 @@ public void ExplicitCast_StaticType()
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         t1.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t1").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void UnconstrainedTypeParameter_Return()
+        {
+            var source =
+@"class C
+{
+    static object? F0<T>(T t) => t;
+    static object F1<T>(T t) => t;
+    static T F2<T>(T t) => t;
+    static T F3<T, U>(U u) where U : T => u;
+    static T F4<T, U>(U u) where U : class, T => u;
+    static T F5<T, U>(U u) where U : struct, T => u;
+    static U F6<T, U>(T t) where U : T => (U)t;
+    static U F7<T, U>(T t) where U : class, T => (U)t;
+    static U F8<T, U>(T t) where U : struct, T => (U)t;
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (4,33): warning CS8603: Possible null reference return.
+                //     static object F1<T>(T t) => t;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "t").WithLocation(4, 33));
+        }
+
+        // PROTOTYPE(NullableReferenceTypes): Track nullability of unconstrained type parameters.
+        [Fact(Skip = "TODO")]
+        public void TrackUnconstrainedTypeParameter_LocalsAndParameters()
+        {
+            var source =
+@"class C
+{
+    static void F1<T>()
+    {
+        default(T).ToString();
+        T x1 = default;
+        x1.ToString();
+        x1!.ToString();
+        if (x1 != null) x1.ToString();
+    }
+    static void F2<T>(T x2)
+    {
+        x2.ToString();
+        x2!.ToString();
+        if (x2 != null) x2.ToString();
+        T y2 = x2;
+        y2.ToString();
+    }
+    static void F3<T>() where T : new()
+    {
+        T x3 = new T();
+        x3.ToString();
+        x3!.ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,9): warning CS8602: Possible dereference of a null reference.
+                //         default(T).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "default(T)").WithLocation(5, 9),
+                // (6,16): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         T x1 = default;
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "default").WithLocation(6, 16),
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         x1.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1").WithLocation(7, 9),
+                // (15,9): warning CS8602: Possible dereference of a null reference.
+                //         x2.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(15, 9));
+        }
     }
```

Comment:
>} [](start = 4, length = 1)

Consider if there are any interesting scenarios when nullable and not nullable value types are converted to unconstraint type parameters

Reply:
Added `// PROTOTYPE` comment.

---
In reply to: [177574874](https://github.com/dotnet/roslyn/pull/25733#discussion_r177574874) [](ancestors = 177574874)